### ISSUE TITLE
Compiler check for the upper bound of condition cases

### ISSUE THREAD

**Issue by asashour** [dart-lang/sdk#47105]
In the below case, since `C` and `D` extend `B`, then the compiler should detect the type and consider it as `B` instead of `A`

```dart
class A {}

class B extends A {
  void b() {}
}

class C extends B {}
class D extends B {}

void f(A a) {
  if (a is C || a is D) {
    a.b(); // <- The method 'b' isn't defined for the type 'A'
  }
  if (a is C) {
    a.b(); // this is fine
  }
  if (a is D) {
    a.b(); // this is fine
  }
}
```

Any hint of where this logic should be implemented is highly appreciated.

**Comment by bwilkerson** [comment]
Yes, that's part of the flow analysis. Flow analysis is specified in the language spec, so any proposal to change it would require a discussion with the language team.

@stereotype441 @leafpetersen

**Comment by stereotype441** [comment]
Thanks for the bug report, @asashour!  You're not the first person to run into this issue.  I'm planning to experiment with some improvements to flow analysis in Q4, and I've filed https://github.com/dart-lang/language/issues/1845 to make sure I don't forget to think about this.
